server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,581 @@
+
+            if unexpected_dirs:
+                hier.add_controller_list(Hierarchy.UNEXPECTED_CONTROLLERS, controller)
+            hier.add_verify_list(controller)
+
+    return 0
+
+
+# User Hierarchy
+def verify_users(hier):
+    """ 'hier' is UserHierarchy class object."""
+    if not Path(hier.path).is_dir():
+        print(
+            f"The setting for USERS in the config file is {hier.path}, but that is"
+            f" not a directory"
+        )
+        return 1
+
+    user_dirs = Path(hier.path).glob("*")
+
+    for user_path in user_dirs:
+        if user_path.is_dir():
+            hier.add_user_dir(user_path)
+        else:
+            hier.add_unexpected_objects(user_path.name)
+
+    return 0
+
+
+def check_and_dump(f, hier, ihier):
+
+    cnt = 0
+    first_check = hier.check_controller()
+    second_check = ihier.check_controller()
+    if first_check or second_check:
+        hier.header(f, "start")
+        if first_check:
+            cnt = hier.dump(f)
+        if second_check:
+            cnt += ihier.dump(f)
+        hier.header(f, "end")
+    return cnt
+
+
+def check_directory_exists(pbdir_path, pbdirname):
+    """ checks the existence of directories """
+    pbdir = Path(pbdir_path)
+    try:
+        pbdir_p = pbdir.resolve(strict=True)
+    except FileNotFoundError:
+        print(f"{_NAME_}: Bad {pbdirname}={pbdir}", file=sys.stderr)
+        return 1
+
+    if not pbdir_p.is_dir():
+        print(f"{_NAME_}: Bad {pbdirname}={pbdir}", file=sys.stderr)
+        return 1
+
+    return 0
+
+
+def main():
+    cfg_name = os.environ.get("_PBENCH_SERVER_CONFIG")
+    if not cfg_name:
+        print(
+            f"{_NAME_}: ERROR: No config file specified; set CONFIG env variable",
+            file=sys.stderr,
+        )
+        return 2
+
+    try:
+        config = PbenchServerConfig(cfg_name)
+    except BadConfig as e:
+        print(f"{_NAME_}: {e} (config file {cfg_name})", file=sys.stderr)
+        return 1
+
+    if check_directory_exists(config.ARCHIVE, "ARCHIVE") > 0:
+        return 1
+
+    if check_directory_exists(config.INCOMING, "INCOMING") > 0:
+        return 1
+
+    if check_directory_exists(config.RESULTS, "RESULTS") > 0:
+        return 1
+
+    if check_directory_exists(config.USERS, "USERS") > 0:
+        return 1
+
+    logger = get_pbench_logger(_NAME_, config)
+    init_db(config, logger)
+
+    ret = 0
+
+    try:
+        Path(config.LOGSDIR, "pbench-audit-server").mkdir()
+    except FileExistsError:
+        # directory already exists, ignore
+        pass
+    except Exception:
+        print("os.mkdir: Unable to create destination directory")
+        return 1
+    logfile = Path(config.LOGSDIR, "pbench-audit-server", "pbench-audit-server.log")
+    # TEMPORARY addition of error file for the sake of test cases
+    errorfile = Path(config.LOGSDIR, "pbench-audit-server", "pbench-audit-server.error")
+    with errorfile.open(mode="w") as f:
+        pass
+    # END
```

Comment:
This code doesn't accomplish anything obvious (perhaps it creates an empty file?...but there are probably better ways to do that...).  It should either be removed or properly commented (what is there isn't adequate, and the `TEMPORARY` label makes me want it removed before merging).

Reply:
yes this is a temporary file which is empty. as it is mentioned there it is just for the sake of test cases. we can remove it before merging.

Reply:
Good.

Just a note:  anything which we "should remove before merging" would be be best removed before reviewing!  😉

Reply:
It will lead to failure in test cases. but I will take care about this next time

Reply:
I'm not sure what you mean by "next time".  If you intend to merge the module with this code in it, then I think the comment needs to be expanded.

And, if all that it is supposed to do is create an empty file, I think we can find a better way to do that ([e.g.](https://docs.python.org/3/library/pathlib.html#pathlib.Path.touch)).

Reply:
By 'next time', I meant next commit.